docs/ServiceWorkerPlugin.md

```diff
@@ -0,0 +1,79 @@
+# ServiceWorkerPlugin
+
+Webpack plugin for configuring a ServiceWorker for different PWA development
+scenarios.
+
+## Purpose
+
+This plugin is a wrapper around the [Google Workbox Webpack Plugin](https://developers.google.com/web/tools/workbox/guides/generate-service-worker/).
+It generates a caching ServiceWorker based on assets emitted by Webpack.
+
+This plugin can be configured to run in the following modes:
+
+- *normal development* - ServiceWorker is disabled
+- *service worker debugging* - ServiceWorker and hot-reloading are enabled
+
+## Usage
+
+In `webpack.config.js`:
+
+```js
+const path = require('path');
+const buildpack = require('@magento/pwa-buildpack');
+const ServiceWorkerPlugin = buildpack.Webpack.ServiceWorkerPlugin;
+
+module.exports = async env => {
+    const config = {
+        /* webpack config, i.e. entry, output, etc. */
+        plugins: [
+            /* other plugins */
+            new ServiceWorkerPlugin({
+                env: {
+                    mode: 'development'
```

Comment:
Is this just `development` or `production`? If so, as a plugin, the code already has access to read webpack 4's `mode` setting (required field in config). So we don't need to manually specify here. Less chance for bugs to creep in from bad configs

Reply:
Believe the config will be available on `compiler.options` in your plugin

Reply:
We're very close - I need to review and merge the open PR from #21 and do some testing.

We don't have to switch to `mode` for now, but I'd like to use a different term in our plugin so we're not overloading what `mode` means in a webpack config (`mode` in webpack 4 only accepts 2 values).

Reply:
👍